grpc/src/test/java/com/linecorp/armeria/server/grpc/GrpcServiceServerTest.java

```diff
@@ -0,0 +1,186 @@
+
+import static com.linecorp.armeria.common.http.HttpSessionProtocols.HTTP;
+import static org.assertj.core.api.Assertions.assertThat;
+
+import org.junit.ClassRule;
+import org.junit.Test;
+
+import com.linecorp.armeria.client.http.HttpClient;
+import com.linecorp.armeria.client.http.HttpClientFactory;
+import com.linecorp.armeria.common.grpc.GrpcSerializationFormats;
+import com.linecorp.armeria.common.http.AggregatedHttpMessage;
+import com.linecorp.armeria.common.http.HttpHeaderNames;
+import com.linecorp.armeria.common.http.HttpHeaders;
+import com.linecorp.armeria.common.http.HttpMethod;
+import com.linecorp.armeria.common.http.HttpStatus;
+import com.linecorp.armeria.grpc.testing.TestServiceGrpc;
+import com.linecorp.armeria.grpc.testing.TestServiceGrpc.TestServiceImplBase;
+import com.linecorp.armeria.internal.grpc.GrpcHeaderNames;
+import com.linecorp.armeria.internal.grpc.GrpcTestUtil;
+import com.linecorp.armeria.server.ServerBuilder;
+import com.linecorp.armeria.testing.server.ServerRule;
+
+import come.linecorp.armeria.grpc.testing.Messages.EchoStatus;
+import come.linecorp.armeria.grpc.testing.Messages.SimpleRequest;
+import come.linecorp.armeria.grpc.testing.Messages.SimpleResponse;
+import come.linecorp.armeria.grpc.testing.Messages.StreamingOutputCallRequest;
+import come.linecorp.armeria.grpc.testing.Messages.StreamingOutputCallResponse;
+import io.grpc.Status;
+import io.grpc.StatusRuntimeException;
+import io.grpc.stub.StreamObserver;
+
+public class GrpcServiceServerTest {
+
+    private static class TestServiceImpl extends TestServiceImplBase {
+
+        @Override
+        public void unaryCall(SimpleRequest request, StreamObserver<SimpleResponse> responseObserver) {
+            if (request.hasResponseStatus()) {
+                throw new StatusRuntimeException(Status.fromCodeValue(request.getResponseStatus().getCode()));
+            }
+            responseObserver.onNext(GrpcTestUtil.RESPONSE_MESSAGE);
+            responseObserver.onCompleted();
+        }
+
+        @Override
+        public void streamingOutputCall(StreamingOutputCallRequest request,
+                                        StreamObserver<StreamingOutputCallResponse> responseObserver) {
+            responseObserver.onNext(
+                    StreamingOutputCallResponse.newBuilder()
+                                               .setPayload(request.getPayload())
+                                               .build());
+            responseObserver.onCompleted();
+        }
+    }
+
+    @ClassRule
+    public static ServerRule server = new ServerRule() {
+        @Override
+        protected void configure(ServerBuilder sb) throws Exception {
+            sb.numWorkers(1);
+            sb.port(0, HTTP);
+
+            sb.serviceUnder("/", new GrpcServiceBuilder()
+                    .addService(new TestServiceImpl())
+                    .build()
+                    .decorate(UnframedGrpcService::new));
```

Comment:
Good point - since it only applies to GrpcService, making the decorator the public API doesn't make much sense. However, since we have a builder, I want to keep all settings there and made it a flag in the builder. It means the builder can't return GrpcService though - it feels a little weird, but I think it only matters in unit testing so maybe it's ok. Let me know what you think though.